src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/groupms2/GroupMS2Task.java

```diff
@@ -0,0 +1,164 @@
+ * USA
+ */
+
+package net.sf.mzmine.modules.peaklistmethods.filtering.groupms2;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import com.google.common.collect.Range;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import net.sf.mzmine.datamodel.Feature;
+import net.sf.mzmine.datamodel.MZmineProject;
+import net.sf.mzmine.datamodel.PeakList;
+import net.sf.mzmine.datamodel.PeakListRow;
+import net.sf.mzmine.datamodel.RawDataFile;
+import net.sf.mzmine.datamodel.Scan;
+import net.sf.mzmine.parameters.ParameterSet;
+import net.sf.mzmine.parameters.parametertypes.tolerances.MZTolerance;
+import net.sf.mzmine.parameters.parametertypes.tolerances.RTTolerance;
+import net.sf.mzmine.taskcontrol.AbstractTask;
+import net.sf.mzmine.taskcontrol.TaskStatus;
+
+/**
+ * Filters out peak list rows.
+ */
+public class GroupMS2Task extends AbstractTask {
+
+  // Logger.
+  private static final Logger LOG = Logger.getLogger(GroupMS2Task.class.getName());
+  // Peak lists.
+  private final MZmineProject project;
+  // Processed rows counter
+  private int processedRows, totalRows;
+  // Parameters.
+  private final ParameterSet parameters;
+  private PeakList list;
+  private RTTolerance rtTol;
+  private MZTolerance mzTol;
+  private boolean limitRTByFeature;
+
+  /**
+   * Create the task.
+   *
+   * @param list peak list to process.
+   * @param parameterSet task parameters.
+   */
+  public GroupMS2Task(final MZmineProject project, final PeakList list,
+      final ParameterSet parameterSet) {
+
+    // Initialize.
+    this.project = project;
+    parameters = parameterSet;
+    rtTol = parameters.getParameter(GroupMS2Parameters.rtTol).getValue();
+    mzTol = parameters.getParameter(GroupMS2Parameters.mzTol).getValue();
+    limitRTByFeature = parameters.getParameter(GroupMS2Parameters.limitRTByFeature).getValue();
+    this.list = list;
+    processedRows = 0;
+    totalRows = 0;
+  }
+
+  @Override
+  public double getFinishedPercentage() {
+
+    return totalRows == 0 ? 0.0 : (double) processedRows / (double) totalRows;
+  }
+
+  @Override
+  public String getTaskDescription() {
+
+    return "Adding all MS2 scans to their features in list " + list.getName();
+  }
+
+  @Override
+  public void run() {
+
+    try {
+      setStatus(TaskStatus.PROCESSING);
+
+      totalRows = list.getNumberOfRows();
+      // for all features
+      for (PeakListRow row : list.getRows()) {
+        for (Feature f : row.getPeaks()) {
+          if (getStatus() == TaskStatus.ERROR)
+            return;
+          if (isCanceled())
+            return;
+
+          RawDataFile raw = f.getDataFile();
+          IntArrayList scans = new IntArrayList();
+          int best = f.getRepresentativeScanNumber();
+          double frt = f.getRT();
+          double fmz = f.getMZ();
+          Range<Double> rtRange = f.getRawDataPointsRTRange();
+          int i = best;
+          // left
+          while (i > 1) {
+            try {

```

Comment:
This kind of ````try {} catch (Exception e) {}```` block looks fishy. What kind of exception are you expecting? Maybe clarify in a comment.